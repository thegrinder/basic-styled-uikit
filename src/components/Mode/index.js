export { default as ModeProvider } from './ModeProvider';
export { default as useMode } from './useMode';
