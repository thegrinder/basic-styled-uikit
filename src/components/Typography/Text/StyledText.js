import styled, { css } from 'styled-components';
import { string, bool } from 'prop-types';

import { getBaseLineHeight, calcLineHeight } from '../typographySelectors';
import {
  getTextSizingStyle,
  getTextColor,
  getTextCommonStyle,
  getTextOpacity,
  getTextColorOnBg,
} from './textSelectors';

const propTypes = {
  color: string.isRequired,
  sizing: string.isRequired,
  marginBottom: bool.isRequired,
  emphasis: string.isRequired,
  ignoreBackground: bool.isRequired,
  bgColor: string,
};

const StyledText = styled.span`
  display: block;
  ${({
    theme,
    color,
    sizing,
    marginBottom,
    emphasis,
    ignoreBackground,
    bgColor,
  }) => {
    const sizingStyle = getTextSizingStyle(theme, sizing);
    return css`
      ${getTextCommonStyle(theme)}
      color: ${
        ignoreBackground
          ? getTextColor(theme, color)
          : getTextColorOnBg(theme, bgColor, color)
      };
      line-height: ${calcLineHeight(sizingStyle?.fontSize)};
      margin-bottom: ${marginBottom ? `${getBaseLineHeight(theme)}rem` : '0'};
      opacity: ${getTextOpacity(theme, emphasis)};
      ${sizingStyle}
    `;
  }}
`;

StyledText.propTypes = propTypes;

export default StyledText;
