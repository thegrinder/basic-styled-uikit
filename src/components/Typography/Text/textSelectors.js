import { getTypography } from '../../../theme/themeSelectors';

export const getTextSizingStyle = (theme, sizing) => {
  const sizingStyle = getTypography(theme).text.sizings[sizing];
  if (!sizingStyle) {
    throw new Error(`There is no ${sizing} size in the <Text /> theme`);
  }
  return sizingStyle;
};

export const getTextColor = (theme, color) => {
  let textColor;

  const baseColors = getTypography(theme).colors;
  if (baseColors && baseColors[color]) {
    textColor = baseColors[color];
  }

  const textColors = getTypography(theme).text.colors;
  if (textColors && textColors[color]) {
    textColor = textColors[color];
  }

  if (!textColor) {
    throw new Error(`There is no ${color} color in the <Text /> theme`);
  }
  return textColor;
};

export const getTextColorOnBg = (theme, bgColor, color) => {
  let textColor;

  const baseColors = getTypography(theme).colors;
  if (baseColors && baseColors[color]) {
    textColor = baseColors[color];
  }

  const baseBgColors = getTypography(theme).onBackground;
  if (baseBgColors && baseBgColors[bgColor] && baseBgColors[bgColor][color]) {
    textColor = baseBgColors[bgColor][color];
  }

  const textTheme = getTypography(theme).text;
  if (
    textTheme &&
    textTheme.onBackground &&
    textTheme.onBackground[bgColor] &&
    textTheme.onBackground[bgColor][color]
  ) {
    textColor = textTheme.onBackground[bgColor][color];
  }

  if (!textColor) {
    throw new Error(
      `There is no ${color} color set for ${bgColor} background color in typography theme`
    );
  }

  return textColor;
};

export const getTextOpacity = (theme, emphasis) => {
  const textOpacity = getTypography(theme).opacities[emphasis];
  if (!textOpacity) {
    throw new Error(`There is no ${emphasis} opacity in the <Text /> theme`);
  }
  return textOpacity;
};

export const getTextCommonStyle = theme => getTypography(theme).text.common;
