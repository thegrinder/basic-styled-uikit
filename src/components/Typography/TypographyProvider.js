import { createProvider } from '../../helpers';

const TypographyProvider = createProvider('typography');

export default TypographyProvider;
