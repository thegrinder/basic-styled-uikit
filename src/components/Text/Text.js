import React from 'react';
import { string, bool, oneOf } from 'prop-types';
import StyledText from './StyledText';

const propTypes = {
  /** determines the size of the text; one of: l, m, s, xs */
  sizing: string,
  /**
   * one of: neutral0, neutral50, neutral100, neutral200, neutral300, neutral400,
   * neutral500, neutral600, neutral700, neutral800, neutral900, neutral1000,
   * primary500, success500, danger500, warning500
   * */
  color: string,
  /** rendered html tag */
  as: oneOf(['span', 'p']),
  /** if true adds a bottom margin according to vertical rhythm */
  marginBottom: bool,
};

const defaultProps = {
  sizing: 'm',
  color: 'default',
  as: 'span',
  marginBottom: false,
};

const Text = ({
  sizing,
  color,
  marginBottom,
  as,
  ...rest
}) => (
    <StyledText
      as={as}
      marginBottom={marginBottom}
      sizing={sizing}
      color={color}
      {...rest}
    />
);

Text.propTypes = propTypes;
Text.defaultProps = defaultProps;

export default Text;
