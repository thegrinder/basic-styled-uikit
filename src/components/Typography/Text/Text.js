import React from 'react';
import { string, bool, oneOf } from 'prop-types';
import StyledText from './StyledText';
import { useColorBoxContext } from '../../ColorBox/context';

const propTypes = {
  /** determines the size of the text; one of: l, m, s, xs */
  sizing: string,
  /** one of: neutral, primary, success, danger, warning */
  color: string,
  /** one of: high, normal, low */
  emphasis: string,
  /** determines whether colors defined for the background are ignored */
  ignoreBackground: bool,
  /** rendered html tag */
  as: oneOf(['span', 'p']),
  /** if true adds a bottom margin according to vertical rhythm */
  marginBottom: bool,
};

const defaultProps = {
  sizing: 'm',
  color: 'neutral',
  ignoreBackground: false,
  as: 'span',
  marginBottom: false,
  emphasis: 'high',
};

const Text = ({
  sizing,
  color,
  ignoreBackground,
  marginBottom,
  emphasis,
  as,
  ...rest
}) => (
  <StyledText
    as={as}
    marginBottom={marginBottom}
    sizing={sizing}
    color={color}
    emphasis={emphasis}
    ignoreBackground={ignoreBackground}
    bgColor={useColorBoxContext()}
    {...rest}
  />
);

Text.propTypes = propTypes;
Text.defaultProps = defaultProps;

export default Text;
