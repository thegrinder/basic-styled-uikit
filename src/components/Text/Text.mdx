---
name: Text
menu: Components
---

import { Playground, Props } from 'docz';
import Text from './Text';
import ColorBox from '../ColorBox/ColorBox';


# Text

<Props of={Text} />


### Text tags

<Playground>
  <Text as="span">Span tag</Text>
  <Text as="p">Paragraph tag</Text>
</Playground>


### Text sizes

<Playground>
  <Text sizing="l">Large</Text>
  <Text sizing="m">Medium</Text>
  <Text sizing="s">Small</Text>
  <Text sizing="xs">Extra small</Text>
</Playground>


### Text colors

<Playground>
  <Text color="primary">primary</Text>
  <Text color="danger">danger</Text>
  <Text color="success">success</Text>
  <Text color="warning">warning</Text>
  <Text color="neutral" emphasis="high">neutral emphasis high</Text>
  <Text color="neutral" emphasis="normal">neutral emphasis normal</Text>
  <Text color="neutral" emphasis="low">neutral emphasis low</Text>

</Playground>


### Text with a bottom margin

<Playground>
  <Text marginBottom={true}>Text with margin</Text>
  <Text marginBottom={true}>Text with margin</Text>
</Playground>