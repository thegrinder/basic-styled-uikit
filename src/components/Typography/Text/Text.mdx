---
name: Text
menu: Components
---

import { Playground, Props } from 'docz';
import Text from './Text';
import ColorBox from '../../ColorBox/ColorBox';


# Text

<Props of={Text} />


### Text tags

<Playground>
  <Text as="span">Span tag</Text>
  <Text as="p">Paragraph tag</Text>
</Playground>


### Text sizes

<Playground>
  <Text sizing="l">Large</Text>
  <Text sizing="m">Medium</Text>
  <Text sizing="s">Small</Text>
  <Text sizing="xs">Extra small</Text>
</Playground>


### Text colors

<Playground>
  <Text color="primary">primary</Text>
  <Text color="danger">danger</Text>
  <Text color="success">success</Text>
  <Text color="warning">warning</Text>
  <Text color="neutral" emphasis="high">neutral emphasis high</Text>
  <Text color="neutral" emphasis="medium">neutral emphasis medium</Text>
  <Text color="neutral" emphasis="low">neutral emphasis low</Text>

</Playground>

### Text colors on different backgrounds

<Playground>
  <ColorBox bgColor="primary">
    <Text>neutral on primary background</Text>
  </ColorBox>
  <ColorBox bgColor="success">
    <Text>neutral on success background</Text>
  </ColorBox>
  <ColorBox bgColor="danger">
    <Text>neutral on danger background</Text>
  </ColorBox>
  <ColorBox bgColor="warning">
    <Text>neutral on warning background</Text>
  </ColorBox>
  <ColorBox bgColor="warning">
    <Text ignoreBackground={true} color="primary">
      background ignored - primary text
    </Text>
  </ColorBox>
</Playground>


### Text with a bottom margin

<Playground>
  <Text marginBottom={true}>Text with margin</Text>
  <Text marginBottom={true}>Text with margin</Text>
</Playground>