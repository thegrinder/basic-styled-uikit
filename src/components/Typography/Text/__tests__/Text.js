import React from 'react';
import { render } from '@testing-library/react';
import { ModeProvider } from '../../../Mode';
import TypographyProvider from '../../TypographyProvider';
import Text from '../Text';

const children = 'children';

const renderComponent = (props = {}) =>
  render(
    <ModeProvider>
      <TypographyProvider>
        <Text {...props}>{children}</Text>
      </TypographyProvider>
    </ModeProvider>
  );

describe('<Text />', () => {
  it('should render correctly with default props', () => {
    const {
      container: { firstChild },
    } = renderComponent();
    expect(firstChild).toBeDefined();
    expect(firstChild).toHaveTextContent(children);
    expect(firstChild).toMatchSnapshot();
  });

  it('should render correctly with custom props', () => {
    const {
      container: { firstChild },
    } = renderComponent({
      sizing: 'xs',
      color: 'primary',
      marginBottom: true,
      ignoreBackground: true,
    });
    expect(firstChild).toBeDefined();
    expect(firstChild).toHaveTextContent(children);
    expect(firstChild).toMatchSnapshot();
  });

  it('should render <span> tag by default', () => {
    const {
      container: { firstChild },
    } = renderComponent();
    expect(firstChild.tagName).toEqual('SPAN');
  });

  it('should render render <p> tag', () => {
    const {
      container: { firstChild },
    } = renderComponent({ as: 'p' });
    expect(firstChild.tagName).toEqual('P');
  });
});
