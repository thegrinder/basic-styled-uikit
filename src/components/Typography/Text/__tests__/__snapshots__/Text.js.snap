// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Text /> should render correctly with custom props 1`] = `
<span
  class="StyledText-c4xq8q-0 dhJwFL"
  color="primary"
>
  children
</span>
`;

exports[`<Text /> should render correctly with default props 1`] = `
<span
  class="StyledText-c4xq8q-0 lNzUf"
  color="neutral"
>
  children
</span>
`;
